src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs

```diff
@@ -407,142 +407,162 @@ internal virtual ImmutableArray<NamedTypeSymbol> GetAdditionalTopLevelTypes()
-                                    {
-                                        // Forwarded type conflicts with exported type
-                                        context.Diagnostics.Add(new CSDiagnostic(new CSDiagnosticInfo(
-                                            ErrorCode.ERR_ForwardedTypeConflictsWithExportedType, exportedType, exportedType.ContainingAssembly, contender, contender.ContainingModule), NoLocation.Singleton));
-                                    }
-                                    else
-                                    {
-                                        // Forwarded type conflicts with another forwarded type
-                                        context.Diagnostics.Add(new CSDiagnostic(new CSDiagnosticInfo(
-                                            ErrorCode.ERR_ForwardedTypesConflict, exportedType, exportedType.ContainingAssembly, contender, contender.ContainingAssembly), NoLocation.Singleton));
-                                    }
-                                }
-
-                                continue;
+                                // Forwarded type conflicts with exported type
+                                diagnostics.Add(new CSDiagnostic(new CSDiagnosticInfo(
+                                    ErrorCode.ERR_ForwardedTypeConflictsWithExportedType, type, type.ContainingAssembly, contender, contender.ContainingModule), NoLocation.Singleton));
+                            }
+                            else
+                            {
+                                // Forwarded type conflicts with another forwarded type
+                                diagnostics.Add(new CSDiagnostic(new CSDiagnosticInfo(
+                                    ErrorCode.ERR_ForwardedTypesConflict, type, type.ContainingAssembly, contender, contender.ContainingAssembly), NoLocation.Singleton));
                             }
-
-                            exportedNamesMap.Add(fullEmittedName, exportedType);
                         }
+
+                        continue;
                     }
+
+                    exportedNamesMap.Add(fullEmittedName, type);
                 }
             }
-
-            return _lazyExportedTypes;
         }
 
         private static void GetForwardedTypes(
             HashSet<NamedTypeSymbol> seenTopLevelTypes,
             CommonAssemblyWellKnownAttributeData<NamedTypeSymbol> wellKnownAttributeData,
-            ArrayBuilder<NamedTypeSymbol> builder)
+            ArrayBuilder<Cci.ExportedType> builder)
         {
-            if (wellKnownAttributeData != null && wellKnownAttributeData.ForwardedTypes != null)
+            if (wellKnownAttributeData?.ForwardedTypes?.Count > 0)
             {
+                // (type, index of the parent exported type in builder, or -1 if the type is a top-level type)
+                var stack = new Stack<ValueTuple<NamedTypeSymbol, int>>();
```

Comment:
Done.

---
In reply to: [67208258](https://github.com/dotnet/roslyn/pull/12004#discussion_r67208258) [](ancestors = 67208258)